mlrun/model_monitoring/applications/base.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2023 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Union, cast
+
+import numpy as np
+import pandas as pd
+
+import mlrun
+import mlrun.model_monitoring.applications.context as mm_context
+import mlrun.model_monitoring.applications.results as mm_results
+from mlrun.serving.utils import StepToDict
+
+
+class ModelMonitoringApplicationBaseV2(StepToDict, ABC):
+    """
+    A base class for a model monitoring application.
+    Inherit from this class to create a custom model monitoring application.
+
+    example for very simple custom application::
+        # mlrun: start-code
+        class MyApp(ApplicationBase):
+            def do_tracking(
+                self,
+                monitoring_context: mm_context.MonitoringApplicationContext,
+            ) -> Union[
+                mm_results.ModelMonitoringApplicationResult,
+                list[
+                    Union[
+                        mm_results.ModelMonitoringApplicationResult,
+                        mm_results.ModelMonitoringApplicationMetric,
+                    ]
+                ],
+                dict[str, Any],
+            ]:
+                monitoring_context.log_artifact(
+                    TableArtifact(
+                        "sample_df_stats", df=self.dict_to_histogram(sample_df_stats)
+                    )
+                )
+                return [
+                    ModelMonitoringApplicationResult(
+                        name="data_drift_test",
+                        value=0.5,
+                        kind=mm_constant.ResultKindApp.data_drift,
+                        status=mm_constant.ResultStatusApp.detected,
+                    )
+                ]
+
+
+        # mlrun: end-code
+    """
+
+    kind = "monitoring_application"
+
+    def do(
```

Comment:
Since this method is essentially private (we do not expect user overrides) but it has to be named `do` - I suggest renaming it to `_do` and adding in the class `do = _do`.
I think that it will hide this method in the API reference, which also needs to be updated: `docs/api/mlrun.model_monitoring.rst`

Reply:
I can also delete the do at all by adding in the class `do = do_tracking` sounds better? because the `do` function isn't doing something important now...

Reply:
lets use the `:exclude-members: do`  in the docs for this